wolfcrypt/src/port/atmel/atmel.c

```diff
@@ -0,0 +1,240 @@
+	}
+    atcab_printbin_label((const uint8_t*)"\r\nRandom Number", rand_out, count);
+#else
+    // TODO: Use on-board TRNG
+#endif
+	return ret;
+}
+
+int atmel_get_random_block(unsigned char* output, unsigned int sz)
+{
+	return atmel_get_random_number((uint32_t)sz, (uint8_t*)output);
+}
+
+extern struct rtc_module *_rtc_instance[RTC_INST_NUM];
+long atmel_get_curr_time_and_date(long* tm)
+{
+    (void)tm;
+
+	/* Get current time */
+    //struct rtc_calendar_time rtcTime;
+	//rtc_calendar_get_time(_rtc_instance[0], &rtcTime);
+
+    /* Convert rtc_calendar_time to seconds since UTC */
+    
+    return 0;
+}
+
+
+
+#ifdef WOLFSSL_ATECC508A
+
+/* Function to allocate new slot number */
+int atmel_ecc_alloc(void)
+{
+    int i, slot = -1;
+    for (i=0; i <= ATECC_MAX_SLOT; i++) {
+        /* Find free slot */
+        if (mSlotList[i] == ATECC_INVALID_SLOT) {
+            mSlotList[i] = i;
+            slot = i;
+            break;
+        }
+    }
+    return slot;
+}
+
+
+/* Function to return slot number to avail list */
+void atmel_ecc_free(int slot)
+{
+    if (slot >= 0 && slot <= ATECC_MAX_SLOT) {
+        /* Mark slot of free */
+        mSlotList[slot] = ATECC_INVALID_SLOT;
+    }
+}
+
+
+/**
+ * \brief Give enc key to read pms.
+ */
+static int atmel_set_enc_key(uint8_t* enckey, int16_t keysize)
+{
+    if (enckey == NULL || keysize != ATECC_KEY_SIZE) {
+        return -1;
+    }
+
+    XMEMSET(enckey, 0xFF, keysize); // use default values
+
+    return SSL_SUCCESS;
+}
+
+
+/**
+ * \brief Write enc key before.
+ */
+static int atmel_init_enc_key(void)
+{
+	uint8_t ret = 0;
+	uint8_t read_key[ATECC_KEY_SIZE] = { 0 };
+
+	XMEMSET(read_key, 0xFF, sizeof(read_key));
+	ret = atcab_write_bytes_slot(0x04, 0, read_key, sizeof(read_key));
+	if (ret != ATCA_SUCCESS) {
+		printf("Failed to write key\r\n");
+		return -1;
+	}
+
+	ret = atcatlsfn_set_get_enckey(atmel_set_enc_key);
+	if (ret != ATCA_SUCCESS) {
+		printf("Failed to set enckey\r\n");
+		return -1;
+	}
+
+	return ret;
+}
+
+#endif /* WOLFSSL_ATECC508A */
+
+
+void atmel_init(void)
+{
+    if (!mAtcaInitDone) {
+#ifdef WOLFSSL_ATECC508A
+        int i;
+    #ifdef GET_REV_INFO
+        uint32_t revision = 0;
+    #endif
+        /* Init the free slot list */
+        for (i=0; i<=ATECC_MAX_SLOT; i++) {
+            if (i == 0 || i == 2 || i == 7) {
+                /* ECC Slots (mark avail) */
+                mSlotList[i] = ATECC_INVALID_SLOT;
+            }
+            else {
+                mSlotList[i] = i;
+            }
+        }
+
+        /* Initialize the CryptoAuthLib to communicate with ATECC508A */
+        atcatls_init(&cfg_ateccx08a_i2c_default);
+
+        /* Init the I2C pipe encryption key. */
+        /* Value is generated/stored during pair for the ATECC508A and stored 
+            on micro flash */
+        /* For this example its a fixed value */
+		if (atmel_init_enc_key() != ATCA_SUCCESS) {
+			printf("Failed to initialize transport key\r\n");
```

Comment:
Can you assume printf() will be available?  Shouldn't we return an error here and log a message instead?